std/node/_stream/async_iterator.ts

```diff
@@ -0,0 +1,260 @@
+function createIterResult(value: any, done: boolean): ReadableIteratorResult {
+  return { value, done };
+}
+
+function readAndResolve(iter: ReadableStreamAsyncIterator) {
+  const resolve = iter[kLastResolve];
+  if (resolve !== null) {
+    const data = iter[kStream].read();
+    // We defer if data is null. We can be expecting either 'end' or 'error'.
+    if (data !== null) {
+      iter[kLastPromise] = null;
+      iter[kLastResolve] = null;
+      iter[kLastReject] = null;
+      resolve(createIterResult(data, false));
+    }
+  }
+}
+
+function onReadable(iter: ReadableStreamAsyncIterator) {
+  queueMicrotask(() => readAndResolve(iter));
+}
+
+function wrapForNext(
+  lastPromise: Promise<ReadableIteratorResult>,
+  iter: ReadableStreamAsyncIterator,
+) {
+  return (
+    resolve: (value: ReadableIteratorResult) => void,
+    reject: (error: Error) => void,
+  ) => {
+    lastPromise.then(() => {
+      if (iter[kEnded]) {
+        resolve(createIterResult(undefined, true));
+        return;
+      }
+
+      iter[kHandlePromise](resolve, reject);
+    }, reject);
+  };
+}
+
+function finish(self: ReadableStreamAsyncIterator, err?: Error) {
+  return new Promise(
+    (
+      resolve: (result: ReadableIteratorResult) => void,
+      reject: (error: Error) => void,
+    ) => {
+      const stream = self[kStream];
+
+      finished(stream, (err) => {
+        if (err && err.code !== "ERR_STREAM_PREMATURE_CLOSE") {
+          reject(err);
+        } else {
+          resolve(createIterResult(undefined, true));
+        }
+      });
+      destroyer(stream, err);
+    },
+  );
+}
+
+const AsyncIteratorPrototype = Object.getPrototypeOf(
+  Object.getPrototypeOf(async function* () {}).prototype,
+);
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+class ReadableStreamAsyncIterator implements AsyncIterableIterator<any> {
+  [kEnded]: boolean;
+  [kError]: Error | null = null;
+  [kHandlePromise] = (
+    resolve: (value: ReadableIteratorResult) => void,
+    reject: (value: Error) => void,
+  ) => {
+    const data = this[kStream].read();
+    if (data) {
+      this[kLastPromise] = null;
+      this[kLastResolve] = null;
+      this[kLastReject] = null;
+      resolve(createIterResult(data, false));
+    } else {
+      this[kLastResolve] = resolve;
+      this[kLastReject] = reject;
+    }
+  };
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [kLastPromise]: null | Promise<any>;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [kLastReject]: null | ((value: any) => void) = null;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [kLastResolve]: null | ((value: any) => void) = null;
+  [kStream]: Readable;
+  [Symbol.asyncIterator] = AsyncIteratorPrototype[Symbol.asyncIterator];
+
+  constructor(stream: Readable) {
+    this[kEnded] = stream.readableEnded || stream._readableState.endEmitted;
+    this[kStream] = stream;
+    Object.defineProperties(this, {
```

Comment:
this could be DRY'ed up quite a bit... a simple helper like:

```ts
function initSymbols(o: any, symbols: symbol[]): any {
  let properties: PropertyDescriptorMap = {};
  for (const sym in symbols) {
    properties[sym] = { configurable: false, enumerable: false, writable: true };
  }
  return Object.defineProperties(o, properties);
}
```

And then all you would need to do is:

```ts
initSymbols(this, [kEnded, kError, kHandlePromise, kLastPromise, kLastReject, kLastResolve, kStream ]);
```

Reply:
Thanks for the code snippet. Implementing right now